Fix crash with opening file with editor on an empty panel

booth-w opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced runtime validations for the superfile process, allowing for more granular control during testing.

- **Tests**
  - Introduced a new automated test class to validate application stability when performing operations on an empty file panel.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lazysegtree commented:
It also crashes on opening Ctrl+A on empty file panel. Can you also fix that

lazysegtree commented:
@booth-w Lets also evaluate all other possible inputs when spf migth crash on empty file panel.

booth-w commented:
Tested in both browser and select mode

open file with xdg-open: working (only tested on Linux but there is a check so it should work)
select all: working
toggle dot files: working
search: fixed
rename: working
copy: working
cut: working
delete: working and removed unnecessary warning when in browser mode
copy path: fixed
extract: fixed
compress: fixed
open with editor: fixed

Sorry this took a bit. I have been busy with uni course work.

lazysegtree commented:
I am gonna test this out on windows, and also will try review the code wherever we access `panel.element` slice.

It might be in other places as well, like process bar's  process list, and sidebar's list.

I plan to get this issue fixed once and for all.

lazysegtree commented:
Going into select mode by pressing "V" and then pressing shift+up still causes crash.

I might write a testsuite for this to validate for no crash on all inputs

lazysegtree commented:
Rebased with main

lazysegtree commented on src/internal/model_render.go:
```diff
@@ -648,7 +650,8 @@ func (m *model) filePreviewPanelRender() string {
 		return box.Render("\n --- " + icon.Error + " No content to preview ---")
 	}
-
+	// This could create errors if panel.cursor ever becomes negative, or goes out of bounds
+	// We should have a panel validation function in our View() function
```
We cant call this function before accessing panel.element[panel.cursor] everytime. 
Instead we would have some kind of validation whenever we update panel.cursor and panel.elements.

lazysegtree replied:
But not in this PR.

lazysegtree replied:
I am not going to do it now. It will be done as the part of refactoring when  we would abstract out all operations to panel objects with struct methods with filePanel recievers.

lazysegtree commented on src/internal/function.go:
```diff
@@ -208,6 +208,10 @@ func returnDirElementBySearchString(location string, displayDotFile bool, search
 		return []element{}
 	}
 
+	if len(items) == 0 {
```
@coderabbitai help 
Can you check if there are any more places where we need to add this kind of check ?
I saw these are the places where panel.elements is checked 
<details><summary>Details</summary>
<p>

```
➜  ~/Workspace/Other_proj/spf_forks/booth-w/superfile/src git:(fix-editor-empty-dir) [3:22:24] grepr "panel.element\["
./internal/model_render.go:191:				isItemSelected := arrayContains(filePanel.selected, filePanel.element[h].location)
./internal/model_render.go:195:					_, err := os.ReadDir(filePanel.element[h].location)
./internal/model_render.go:196:					f[i] += filePanelCursorStyle.Render(cursor+" ") + prettierName(filePanel.element[h].name, m.fileModel.width-5, filePanel.element[h].directory || (err == nil), isItemSelected, filePanelBGColor) + endl
./internal/model_render.go:655:	itemPath := panel.element[panel.cursor].location
./internal/model.go:463:			if filepath.Dir(focusPanel.element[0].location) != focusPanel.location {
./internal/function.go:368:	if len(panel.element[panel.cursor].metaData) != 0 && m.focusPanel != metadataFocus {
./internal/function.go:369:		m.fileMetaData.metaData = panel.element[panel.cursor].metaData
./internal/function.go:374:	filePath := panel.element[panel.cursor].location
./internal/function.go:451:	panel.element[panel.cursor].metaData = m.fileMetaData.metaData
./internal/handle_modal.go:73:	oldPath := panel.element[panel.cursor].location
./internal/handle_panel_up_down.go:112:	if arrayContains(panel.selected, panel.element[selectItemIndex].location) {
./internal/handle_panel_up_down.go:113:		panel.selected = removeElementByValue(panel.selected, panel.element[selectItemIndex].location)
./internal/handle_panel_up_down.go:115:		panel.selected = append(panel.selected, panel.element[selectItemIndex].location)
./internal/handle_panel_up_down.go:135:	if arrayContains(panel.selected, panel.element[selectItemIndex].location) {
./internal/handle_panel_up_down.go:136:		panel.selected = removeElementByValue(panel.selected, panel.element[selectItemIndex].location)
./internal/handle_panel_up_down.go:138:		panel.selected = append(panel.selected, panel.element[selectItemIndex].location)
./internal/handle_panel_movement.go:52:	if panel.element[panel.cursor].directory {
./internal/handle_panel_movement.go:57:		panel.location = panel.element[panel.cursor].location
./internal/handle_panel_movement.go:67:	} else if !panel.element[panel.cursor].directory {
./internal/handle_panel_movement.go:68:		fileInfo, err := os.Lstat(panel.element[panel.cursor].location)
./internal/handle_panel_movement.go:75:			targetPath, symlink_err := filepath.EvalSymlinks(panel.element[panel.cursor].location)
./internal/handle_panel_movement.go:101:			cmd := exec.Command(dllpath, dllfile, panel.element[panel.cursor].location)
./internal/handle_panel_movement.go:110:		cmd := exec.Command(openCommand, panel.element[panel.cursor].location)
./internal/handle_panel_movement.go:160:		elementLocation := panel.element[panel.cursor].location
./internal/handle_file_operations.go:45:	oldPath := panel.element[panel.cursor].location
./internal/handle_file_operations.go:79:	cursorPos := strings.LastIndex(panel.element[panel.cursor].name, ".")
./internal/handle_file_operations.go:80:	nameLen := len(panel.element[panel.cursor].name)
./internal/handle_file_operations.go:81:	if cursorPos == -1 || cursorPos == 0 && nameLen > 0 || panel.element[panel.cursor].directory {
./internal/handle_file_operations.go:93:	ti.SetValue(panel.element[panel.cursor].name)
./internal/handle_file_operations.go:152:		name:     icon.Delete + icon.Space + panel.element[panel.cursor].name,
./internal/handle_file_operations.go:167:	err := trashMacOrLinux(panel.element[panel.cursor].location)
./internal/handle_file_operations.go:269:		name:     "󰆴 " + panel.element[panel.cursor].name,
./internal/handle_file_operations.go:285:	err := os.RemoveAll(panel.element[panel.cursor].location)
./internal/handle_file_operations.go:388:		"panel location", panel.element[panel.cursor].location)
./internal/handle_file_operations.go:389:	m.copyItems.items = append(m.copyItems.items, panel.element[panel.cursor].location)
./internal/handle_file_operations.go:530:	ext := strings.ToLower(filepath.Ext(panel.element[panel.cursor].location))
./internal/handle_file_operations.go:536:	outputDir := fileNameWithoutExtension(panel.element[panel.cursor].location)
./internal/handle_file_operations.go:548:	err = extractCompressFile(panel.element[panel.cursor].location, outputDir)
./internal/handle_file_operations.go:563:	fileName := filepath.Base(panel.element[panel.cursor].location)
./internal/handle_file_operations.go:573:	err = zipSource(panel.element[panel.cursor].location, filepath.Join(filepath.Dir(panel.element[panel.cursor].location), zipName))
./internal/handle_file_operations.go:609:	args := append(parts[1:], panel.element[panel.cursor].location)
./internal/handle_file_operations.go:653:	if err := clipboard.WriteAll(panel.element[panel.cursor].location); err != nil {
➜  ~/Workspace/Other_proj/spf_forks/booth-w/superfile/src git:(fix-editor-empty-dir) [6:05:34]
```



</p>
</details>

lazysegtree replied:
Thanks. Fixed it.

lazysegtree commented:
@coderabbitai can you gather the PR count of this repo and render that as a line chart, showing count of open/closed/total PRs over time.

lazysegtree commented:
@coderabbitai full review

lazysegtree commented:
@booth-w I am good with the PR now. Please review and approve once

booth-w commented:
What do you want me to do? I don't think I can approve my own PR.

lazysegtree commented:
Testsuite was not passing earlier but passes now 

```
➜  ~/Workspace/Other_proj/spf_forks/booth-w/superfile git:(fix-editor-empty-dir) [7:51:29] $SPF_SRC/testsuite/.venv/bin/python3 testsuite/main.py -d -t EmptyPanelTest
[2025-03-29 19:51:37 -   DEBUG] server object : Server(socket_name=superfile)
[2025-03-29 19:51:37 -   DEBUG] Initialized TestFSManager
[2025-03-29 19:51:37 -   DEBUG] Found a testcase EmptyPanelTest, in module empty_panel_test
[2025-03-29 19:51:37 -    INFO] Testcases : [EmptyPanelTest]
[2025-03-29 19:51:37 -    INFO] Running test EmptyPanelTest
[2025-03-29 19:51:37 -   DEBUG] Current file structure :
D-dir1

[2025-03-29 19:51:38 -   DEBUG] spf_session initialised : Session($0 spf_session)
[2025-03-29 19:51:39 -   DEBUG] spf_manager info : [Session($0 spf_session)], Current file structure :
D-dir1

[2025-03-29 19:51:39 -    INFO] Passed test EmptyPanelTest
[2025-03-29 19:51:39 -    INFO] Finished running 1 test. 1 passed
➜  ~/Workspace/Other_proj/spf_forks/booth-w/superfile git:(fix-editor-empty-dir) [7:51:39] git checkout yorukot/main -- src

➜  ~/Workspace/Other_proj/spf_forks/booth-w/superfile git:(fix-editor-empty-dir) ✗ [7:51:48] ./build.sh; bin/spf
➜  ~/Workspace/Other_proj/spf_forks/booth-w/superfile git:(fix-editor-empty-dir) ✗ [7:52:01] $SPF_SRC/testsuite/.venv/bin/python3 testsuite/main.py -d -t EmptyPanelTest
[2025-03-29 19:52:02 -   DEBUG] server object : Server(socket_name=superfile)
[2025-03-29 19:52:02 -   DEBUG] Initialized TestFSManager
[2025-03-29 19:52:02 -   DEBUG] Found a testcase EmptyPanelTest, in module empty_panel_test
[2025-03-29 19:52:02 -    INFO] Testcases : [EmptyPanelTest]
[2025-03-29 19:52:02 -    INFO] Running test EmptyPanelTest
[2025-03-29 19:52:02 -   DEBUG] Current file structure :
D-dir1

[2025-03-29 19:52:03 -   DEBUG] spf_session initialised : Session($0 spf_session)
[2025-03-29 19:52:04 -   DEBUG] spf_manager info : [], Current file structure :
D-dir1

[2025-03-29 19:52:04 -   DEBUG] Test assertion failed : Superfile is not running
Traceback (most recent call last):
  File "/Users/kuknitin/Workspace/Other_proj/spf_forks/booth-w/superfile/testsuite/core/base_test.py", line 117, in validate
    assert self.env.spf_mgr.is_spf_running(), "Superfile is not running"
AssertionError: Superfile is not running
[2025-03-29 19:52:04 -   ERROR] Failed test EmptyPanelTest
[2025-03-29 19:52:04 -    INFO] Finished running 1 test. 0 passed
➜  ~/Workspace/Other_proj/spf_forks/booth-w/superfile git:(fix-editor-empty-dir) ✗ [7:52:04]
```

lazysegtree commented:
@booth-w Ohh, right. 
Just wanted you to go through the new changes, and see if everything looks good.

lazysegtree commented:
My bad.. I force pushed the wrong commit.

lazysegtree commented:
Fixing it now

lazysegtree commented:
@booth-w PR is ready now. You can give a go ahead, and I will merge it.

lazysegtree commented:
We just added coderabbit, so I was playing around with it for a while.

booth-w commented:
LGTM, though my only experience with Go is this project so if there is something you have missed, I doubt I will be able to catch it.

Is coderabbit the best for AI code review? I know nothing about it or any others so it could be the best one but the last commit was two years ago and it was archived last year

lazysegtree commented:
@booth-w We are not using coderabbit open source version. We are using coderabbit pro for now.

<img width="765" alt="image" src="https://github.com/user-attachments/assets/bd6de044-a8b2-45f5-9c41-809942d0ed37" />

lazysegtree commented:
Thanks @booth-w for identifying and fixing this. This is a major improvement.
